hack/release/release.go

```diff
@@ -0,0 +1,444 @@
+		"calicoctl",
+		"app-policy",
+		"typha",
+		"felix",
+		"node",
+		"calico", // Technically not a container image, but a helm chart.
+	}
+
+	// Build env.
+	// TODO: Pass CHART_RELEASE to calico repo if needed.
+	env := append(os.Environ(),
+		fmt.Sprintf("VERSION=%s", ver),
+		fmt.Sprintf("DEV_REGISTRIES=%s", strings.Join(registries, " ")),
+		fmt.Sprintf("VALIDARCHES=%s", strings.Join(architectures, " ")),
+	)
+
+	for _, dir := range releaseDirs {
+		err := makeInDirectory(dir, "release-build", env...)
+		if err != nil {
+			return fmt.Errorf("Failed to build %s: %s", dir, err)
+		}
+	}
+	return nil
+}
+
+func publishGithubRelease(ver string) error {
+	releaseNoteTemplate := `
+Release notes can be found at https://projectcalico.docs.tigera.io/archive/{release_stream}/release-notes/
+
+Attached to this release are the following artifacts:
+
+- {release_tar}: container images, binaries, and kubernetes manifests.
+- {calico_windows_zip}: Calico for Windows.
+- {helm_chart}: Calico Helm v3 chart.
+`
+	sv, err := semver.NewVersion(strings.TrimPrefix(ver, "v"))
+	if err != nil {
+		return err
+	}
+	formatters := []string{
+		// Alternating placeholder / filler. We can't use backticks in the multiline string above,
+		// so we replace anything that needs to be backticked into it here.
+		"{version}", ver,
+		"{release_stream}", fmt.Sprintf("v%d.%d", sv.Major, sv.Minor),
+		"{release_tar}", fmt.Sprintf("`release-%s.tgz`", ver),
+		"{calico_windows_zip}", fmt.Sprintf("`calico-windows-%s.zip`", ver),
+		"{helm_chart}", fmt.Sprintf("`tigera-operator-%s.tgz`", ver),
+	}
+	r := strings.NewReplacer(formatters...)
+	releaseNote := r.Replace(releaseNoteTemplate)
+
+	args := []string{
+		"-username", organization,
+		"-repository", repo,
+		"-name", ver,
+		"-body", releaseNote,
+		ver,
+		uploadDir(ver),
+	}
+	_, err = runCommand("ghr", args, nil)
+	return err
+}
+
+func publishContainerImages(ver string) error {
+	releaseDirs := []string{
+		"pod2daemon",
+		"cni-plugin",
+		"apiserver",
+		"kube-controllers",
+		"calicoctl",
+		"app-policy",
+		"typha",
+		"node",
+	}
+
+	env := append(os.Environ(),
+		fmt.Sprintf("IMAGETAG=%s", ver),
+		fmt.Sprintf("VERSION=%s", ver),
+		"RELEASE=true",
+		"CONFIRM=",    // Undo this when done prototyping.
+		"DRYRUN=true", // Undo this when done prototyping.
+		fmt.Sprintf("DEV_REGISTRIES=%s", strings.Join(registries, " ")),
+		fmt.Sprintf("VALIDARCHES=%s", strings.Join(architectures, " ")),
+	)
+
+	for _, dir := range releaseDirs {
+		out, err := makeInDirectoryWithOutput(dir, "release-publish", env...)
+		if err != nil {
+			return fmt.Errorf("Failed to publish %s: %s", dir, err)
+		}
+		logrus.Info(out)
+	}
+	return nil
+}
+
+// determineReleaseVersion uses historical clues to figure out the next semver
+// release number to use for this release.
+func determineReleaseVersion() string {
+	// Check that we're not already on a git tag.
+	_, err := git("describe", "--exact-match", "--tags", "HEAD")
```

Comment:
Worth checking that the repo is up to date and not shallow?  On semaphore that'll be a given but if we run a local release for some reason, it'd be easy to be out of date.